### ISSUE TITLE
Add Hooks for bot events

### ISSUE THREAD

**Issue by deathbybandaid** [sopel-irc/sopel#1546]
I didn't see an open issue/PR for this, but I know Exirel and I have talked about it in IRC.

This would likely be an addition to module.events, with bot events logging to it.

Examples:
* a module.event that waits until the bot is in at least one channel.
* a module that waits for the bot to be "fully loaded" to display a "Bot Loaded" message
    * would also be a good PR for a config option for the bot to announce to `bot.channels.keys()`, default `None`
        * possibly allow `$module_count` and other information to be displayed (now I'm ranting)

I would also suggest this being used as a resource for module dependencies.
`setup(bot)` could be utilized to register that a module is loaded. Other modules could then use that information.

**Comment by deathbybandaid** [comment]
Similar to these

````
RPL_ENDOFMONLIST = '733'
ERR_MONLISTFULL = '734'
````

bot events could be `BOT_xxxxx`

and module events could be `MOD_xxxxx`

**Comment by dgw** [comment]
Sopel's hooks **need** to go somewhere besides `tools.events`. Its docstring even says (emphasis mine):

> An enumeration of all the **standardized** and notable IRC numeric events

Sopel-specific hooks don't belong in that list. We could end up with conflicting numerics as the IRC protocol continues to evolve, and we should make absolutely sure not to have that problem. Yes, we could do something silly like use negative numbers for internal numerics, but… meh.

There's no issue with a `tools.hooks` list to go along with that, with its own namespace, and a corresponding `module.hook` decorator. Hooks don't even have to map to numeric values: we can use strings if we want! (See e.g. [the YOURLS hook system](https://yourls.org/hooklist.php).) I just _really_ don't want to pollute the existing "events" logic with anything not defined in an IRC spec. (Existing IRCd-specific numerics that might exist in `tools.events` notwithstanding, because I didn't add them.)

**Comment by deathbybandaid** [comment]
Mainly, it's the idea that counts

Where it goes exactly, you are right,, should be something seperate from that

**Comment by Exirel** [comment]
> Sopel-specific hooks don't belong in that list.

I fully agree. Hooks are something I really wish to have in Sopel, if not in 7.x, maybe in 8.x. We'll see what we can do.

**Comment by dgw** [comment]
> if not in 7.x, maybe in 8.x

Likely not in 7.x, since we all basically forgot about this idea until ~3 weeks before the target release date of 7.0. 😹 I shoved it in 8.0 so we can get going on it early.

**Comment by deathbybandaid** [comment]
I don't condone my temporary solution for anybody else to use but myself, but to achieve my goals, I made a hacky event dispatcher to trigger a few bot events.

One of the main things I needed was to know that the bot was JOINned to all the channels

**Comment by dgw** [comment]
Roadmap TBD, but not rejected.
